### DISCUSSION TITLE
How to format Markdown with CSS classes?

### DISCUSSION THREAD

**Discussion by BsNoSi** [datenstrom/community#107] [Ask a question] [answered]
Hi,

I am trying to go into deep with Yellow and created my first own plugin for adding style class into generated code. My PHP knowledge is „beginner status“, so I currently have no idea, how to solve it "nice".

I want to add formating into my parsed markdown. Therefore I created a plugin (to be true: modified an existing one) with this main function:

``` .php

function onParseContentBlock($page, $name, $text, $shortcut)
    {
        $output = NULL;
        if($name=="format" && $shortcut)
        {
            list($class, $style) = $this->yellow->toolbox->getTextArgs($text);
      if($class == "end" || $class == "/")
       {
       $output = "<div></div></div>";  
//     $output = "</div>";  // creates wrong html code, "<p></div></p>"
     }
      else
      {
        $output = "<div class=\"".htmlspecialchars($class)."\"";
              if($style) $output .= " style=\"".htmlspecialchars($style)."\"";
              $output .= ">";
      }
        }
        return $output;
    }

```

Starting the div with `[format classname]` creates a line as expected:

``` .html

<div class="classname">

```

To achieve valid html, the closing parameter with `[format end]` or `[format /]` requires as $output

``` .html

<div></div></div>

```

because a simple `</div>` is parsed into `<p></div></p>` in the final output.

For security reasons, I want to keep "parserSafeMode = 1" , so option `markdown="1"` in the output of opening tag does not heal this, if it is not set to "0".

In general, the code delievers, what I want. But I want to understand, what is happening there. Any hints to improve this welcome.

NoSi

**Comment by markseuffert** [comment]
I assume it's happening in the Markdown plugin. Can you upload a working project to reproduce it?

**Comment by BsNoSi** [comment]
I will prepare something and send a link to you.

**Comment by BsNoSi** [comment]
You can [download a zip-file](http://kd.nosi.de/format.zip) for testing. ZIP-File contains working php-file and readme, how to use.

**Comment by markseuffert** [comment]
I suggested support for CSS classes a while ago, see michelf/php-markdown#157. 

Not sure what's the best way to implement it with a plugin. It would be great if it could work without an ending marker `[format classname]`...`[format end]`. One alternative could be `{.classname}`, which then assigns a class to the next upcoming element – whatever that might be.

**Comment by BsNoSi** [comment]
I designed this, to get more than one paragraph into one div for creating sporadically needed columns of different size. [format end] marks clearly the area, that is enclosed in a column.

**Comment by markseuffert** [comment]
Okay, after some experimentation I'd like to suggest three alternatives for `<div>` with CSS classes. 

The first alternative is the one that you suggested, a shortcut with block ending marker. The second alternative is a style block, which can contain Markdown. The third alternative is an HTML block.

First a shortcut to create a `<div class="custom">`:

```
[div custom]
-   First column with two paragraphs. Lorem ipsum dolor sit amet, 
    consectetur adipisicing elit, sed do eiusmod tempor incididunt 
    ut labore et dolore magna pizza.

    Ubicumque homo est, ibi benefici locus est.

-   Second column, just one paragraph.

-   Third column has an image and a list.

    [image icon.png]

    1. first list item
    2. second list item
[/div]
```

Second a style block to create a `<div class="custom">`:

```
::: custom
-   First column with two paragraphs. Lorem ipsum dolor sit amet, 
    consectetur adipisicing elit, sed do eiusmod tempor incididunt 
    ut labore et dolore magna pizza.

    Ubicumque homo est, ibi benefici locus est.

-   Second column, just one paragraph.

-   Third column has an image and a list.

    [image icon.png]

    1. first list item
    2. second list item
:::
```

Third a HTML block to create a single `<div class="custom">`:

```
<div class="custom">
-   First column with two paragraphs. Lorem ipsum dolor sit amet, 
    consectetur adipisicing elit, sed do eiusmod tempor incididunt 
    ut labore et dolore magna pizza.

    Ubicumque homo est, ibi benefici locus est.

-   Second column, just one paragraph.

-   Third column has an image and a list.

    [image icon.png]

    1. first list item
    2. second list item
</div>
```

The first syntax is already possible with our Markdown plugin. You can get rid of the extra wrapper that you mentioned. Please see `function _doAutoLinks_shortcut_callback()` and handle `<div>` and `</div>` as block elements. 

The second syntax is not possible with the current Markdown library. What's interesting about this approach, it's similar to the well-known fenced code block. There's a feature request at michelf/php-markdown#191.

The third syntax is not possible with the current Markdown library. You have to write something like `<div class="columns-block" markdown="1">...</div>`. I think other Markdown libraries allow Markdown inside certain HTML elements, for example CommonMark.

Edited: All alternatives in one post.

**Comment by markseuffert** [comment]
Unfortunately most alternatives are still not possible with the current Markdown library. As soon as features are added there, they become available in Datenstrom Yellow as well.

**Comment by markseuffert** [comment, marked as answer]
I keep coming back to this problem and updated the [Markdown extension](https://github.com/annaesvensson/yellow-markdown).

Start each line with `!` and add any CSS class to the first line:

```
! {.custom}
! Here's a paragraph with custom CSS class.
! **Text with Markdown** can be used optionally.
! [Link to website](https://datenstrom.se).
```

...this will generate the following HTML, nicely wrapped inside a `<div>...</div>`:

``` html
<div class="custom">
<p>Here's a paragraph with custom CSS class.
<strong>Text with Markdown</strong> can be used optionally.
<a href="https://datenstrom.se">Link to website</a>.</p>
</div>
```

For now it's an experimental feature. Let me know how this works for you.

**Comment by pbuzdin** [comment]
Neat!
Thank you!

**Comment by GiovanniSalmeri** [comment]
Just for the record: I checked the relevant discussions at Commonmark:

https://talk.commonmark.org/t/block-directives/802/69
https://talk.commonmark.org/t/generic-directives-plugins-syntax/444/158
https://talk.commonmark.org/t/consistent-attribute-syntax/272/86

... and they seem stalled without any consensus about a syntax. @markseu Your solution seems to me very good. Perhaps I would have a slight preference for the syntax `::: custom .... :::`, which would allow to apply a class to a block of text without modifying it by prefixing each line with a `!`, but definitively your solution works well and is easier to read (the block to which the class is applied is easier to recognise). Thank you!

**Comment by BsNoSi** [comment]
@GiovanniSalmeri : Thanks for the links – I haven't had the idea to check if there is an "offical discussion" about formating blocks. I share your opition `::: custom` block as shown [at the end of this demo](https://markdown-it.github.io/).

**Comment by markseuffert** [comment]
Thanks for your feedback. I don't have a strong opinion what the best syntax is, at the end of the day we need something. If there's consensus what the Markdown syntax should be, we simply update the Markdown extension. The following is supported in Datenstrom Yellow 0.8.5:

* simple notices, with predefined classes in the official themes
* custom notices, `<div>` with custom class or ID
* complex notices, `<div>` can be nested if necessary

Here are some examples, from simple to complex:

```
! Here's a notice with warning

!! Here's a notice with error

!!! Here's a notice with tip
```

```
! {.class}
! Here's a general block element.
! Text can span over several lines
! and contain Markdown text formatting.
```

```
! {.class1}
!! {.class2}
!! Here's a nested `<div>` and
!! it can span over several lines.
```

**Comment by markseuffert** [comment]
I was thinking about the [Markdown syntax](https://github.com/annaesvensson/yellow-markdown) and if it could be simplified a little more.

First, should we rename the term "notice block" to something else like "design block"? It is such a common task for web developers to add `<div>...</div>` to a web page and our editor has supported this for many years. But not sure many people know that. The feature is currently hidden in the text formatting menu under the unusual term `Notice` and is not easy to find unless you already know what to look for.

Second, should we reduce or remove "simple notices", those without any specified CSS class? I can never remember in which order the predefined CSS classes are. At the moment there are three in the official themes. I just use the first one or specify the CSS class by name. It looks like GitHub Flavored Markdown went down a similar route with what they call "alerts",  Obsidian calls them "callouts", you specify by name what kind of alert you want.

Here are simple examples:

```
! Here's a general block element.
! Text can span over several lines
! and contain Markdown text formatting.

! {.important}
! Here's information that needs attention.
```

What do others think, any thoughts from your experience?

**Reply by BsNoSi** [reply]
Hello Mark!

I said goodbye to Yellow quite some time ago because the way the software developed contradicted my idea of sustainable product maintenance. That said, I still think a lot of what you have achieved there is great, and I have incorporated it, at least conceptually, into my solution (a Windows program that generates “finished websites” → ose.nosi.de, only German available).

Perhaps I can give something back, at least in a small way.

I solved the “DIV problem” with double brackets that have to be placed at the beginning of a line:

~~~
(( class class  …
 …
)) optional comment allowed
~~~

This makes even nesting possible because opening/closing is clearly defined:

~~~
(( open1
 …
(( open2
…
)) close open2
))  close open1
~~~

Works fine for me.

**Reply by schulle4u** [reply]
I agree. So far I have hardly used the various notice blocks, only creating them to test themes. I only need the first one for my extensions. However, based on my experience, I find well-named utility classes easier to understand than the different exclamation marks, so my upvote for reducing them. This should also make it more convenient for theme designers to port existing themes to Yellow. "Design Block" is a good name.

**Reply by markseuffert** [reply]
If you don’t mind I like to implement this soon. I did usability testing and it appears to be more intuitive. How about naming it "general block element" in the technical documentation and in the web interface show the blocks which are available, for example "Note" and "Important"? Do you think we need more than two predefined CSS classes in the official themes?

**Reply by schulle4u** [reply]
> How about naming it "general block element" in the technical documentation 

👍

> and in the web interface show the blocks which are available, for example "Note" and "Important"? 

You mean we could give the block elements a specific naming convention so that the available block classes are automatically displayed in the editor menu? 

> Do you think we need more than two predefined CSS classes in the official themes?

No, I think two block classes should be enough as a showcase in the standard themes.

**Reply by markseuffert** [reply]
Please note that the Markdown syntax has changed. The following syntax has replaced notice blocks:

```
! {.important}
! Here's information that needs attention.
```

**Reply by markseuffert** [reply]
> You mean [...] available block classes are automatically displayed in the editor menu?

No, maybe an idea for a later version. For now there's `Important text` in the editor menu.

**Reply by v-bulynkin** [reply]
Please, return three blocks - info, warning and error.
I used all of them, and after update all blocks were gone and, to cap it all, warning and error fully disappeared.
You can change syntax, if you want, but keep all three kinds of blocks, at least for compatibility.
Which necessity forced you to remove them?

**Reply by markseuffert** [reply]
Here are my thoughts. There is still no consensus of what the syntax/name of Markdown block elements should be. Obsidian has more than 10, GitHub has 5 different ones. In Datenstrom Yellow the available CSS classes depend on the theme. In the official themes there is now the `.important` CSS class, a block element with the main colour of a theme. You can customise this and add as many colourful blocks as you need.

However the real reasons behind these changes were my observations in real life. Most websites didn't use all of the predefined blocks and users I watched were overloaded by the many choices in their user interface. As usual I want to write about simplicity in product design, but, I have done that so often. Instead here's what I wrote in the Markdown documentation about block elements:

If you are a web developer you are probably asking, does this mean I can add `<div>...</div>` to a web page and where have you been all my life? The answer is yes and like a diamond in the ground it was waiting to be found.
